go-controller/pkg/factory/factory.go

```diff
@@ -41,25 +41,47 @@ func (factory *Factory) CreateOvnController() *ovn.Controller {
 	namespaceInformer := factory.IFactory.Core().V1().Namespaces()
 
 	return &ovn.Controller{
-		StartPodWatch: func(handler cache.ResourceEventHandler) {
-			podInformer.Informer().AddEventHandler(handler)
+		StartPodWatch: func(handler cache.ResourceEventHandler, processExisting func([]interface{})) {
 			go podInformer.Informer().Run(utilwait.NeverStop)
+			cache.WaitForCacheSync(utilwait.NeverStop, podInformer.Informer().HasSynced)
+			if processExisting != nil {
+				// cache has synced, lets process the list
+				processExisting(podInformer.Informer().GetStore().List())
+			}
+			// now register the event handler
+			podInformer.Informer().AddEventHandler(handler)
 		},
-		StartServiceWatch: func(handler cache.ResourceEventHandler) {
-			serviceInformer.Informer().AddEventHandler(handler)
+		StartServiceWatch: func(handler cache.ResourceEventHandler, processExisting func([]interface{})) {
 			go serviceInformer.Informer().Run(utilwait.NeverStop)
+			cache.WaitForCacheSync(utilwait.NeverStop, serviceInformer.Informer().HasSynced)
+			if processExisting != nil {
+				processExisting(serviceInformer.Informer().GetStore().List())
+			}
+			serviceInformer.Informer().AddEventHandler(handler)
 		},
-		StartEndpointWatch: func(handler cache.ResourceEventHandler) {
-			endpointsInformer.Informer().AddEventHandler(handler)
+		StartEndpointWatch: func(handler cache.ResourceEventHandler, processExisting func([]interface{})) {
 			go endpointsInformer.Informer().Run(utilwait.NeverStop)
+			cache.WaitForCacheSync(utilwait.NeverStop, endpointsInformer.Informer().HasSynced)
+			if processExisting != nil {
+				processExisting(endpointsInformer.Informer().GetStore().List())
+			}
+			endpointsInformer.Informer().AddEventHandler(handler)
 		},
-		StartPolicyWatch: func(handler cache.ResourceEventHandler) {
-			policyInformer.Informer().AddEventHandler(handler)
+		StartPolicyWatch: func(handler cache.ResourceEventHandler, processExisting func([]interface{})) {
 			go policyInformer.Informer().Run(utilwait.NeverStop)
+			cache.WaitForCacheSync(utilwait.NeverStop, policyInformer.Informer().HasSynced)
+			if processExisting != nil {
+				processExisting(policyInformer.Informer().GetStore().List())
+			}
+			policyInformer.Informer().AddEventHandler(handler)
 		},
-		StartNamespaceWatch: func(handler cache.ResourceEventHandler) {
-			namespaceInformer.Informer().AddEventHandler(handler)
+		StartNamespaceWatch: func(handler cache.ResourceEventHandler, processExisting func([]interface{})) {
 			go namespaceInformer.Informer().Run(utilwait.NeverStop)
+			cache.WaitForCacheSync(utilwait.NeverStop, namespaceInformer.Informer().HasSynced)
+			if processExisting != nil {
+				processExisting(namespaceInformer.Informer().GetStore().List())
+			}
+			namespaceInformer.Informer().AddEventHandler(handler)
```

Comment:
Looking at go-controller/pkg/factory/factory.go and go-controller/pkg/ovn/ovn.go, we could reduce the code duplication in pod/endpoint/service/namespace informer watches. I think that can be handled as a separate pr.